# Question 2. KNN
## [CM4] Classify data using a KNN classifier, k-fold cross validation and train-validate-test 

Importing all necessary libraries.

In [35]:
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
from sklearn.model_selection import cross_val_score
import matplotlib.pyplot as plt

# IRIS Data

## Splitting the data for training and testing


In [36]:
df_iris = pd.read_csv("cleaned_data_iris.csv")
x = df_iris.iloc[:,:-1]
y = df_iris.iloc[:,-1]

In [37]:
#Dividing the data into a training set and a test set (80%, 20%)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=98)

In [38]:
# Performing KNN by using default parameters on iris data
knn_default_iris = KNeighborsClassifier()

In [39]:
knn_default_iris.fit(x_train,y_train)

KNeighborsClassifier()

In [40]:
pred=knn_default_iris.predict(x_test)

In [41]:
print(classification_report(y_test,pred))

                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        12
Iris-versicolor       1.00      1.00      1.00         6
 Iris-virginica       1.00      1.00      1.00         3

       accuracy                           1.00        21
      macro avg       1.00      1.00      1.00        21
   weighted avg       1.00      1.00      1.00        21



In [42]:
print(confusion_matrix(y_test,pred))

[[12  0  0]
 [ 0  6  0]
 [ 0  0  3]]


In [43]:
accuracy_score(y_test,pred)

1.0

#### The accuracy of iris dataset using default parameters is 100%

# K fold cross validation for Iris data (K = 5)

In [44]:
from sklearn.model_selection import KFold
kf = KFold(n_splits=5)
accuracy_kfold_iris=[]
for train, test in kf.split(x):
    x1_train,x1_test,y1_train,y1_test=x.iloc[train],x.iloc[test],y.iloc[train],y.loc[test]
    kfold_knn=KNeighborsClassifier()
    kfold_knn.fit(x1_train,y1_train)
    accuracy_kfold_iris.append(accuracy_score(kfold_knn.predict(x1_test),y1_test))
average_accuracy = np.mean(accuracy_kfold_iris)
print("Average accuracy:",average_accuracy)
%store accuracy_kfold_iris

Average accuracy: 0.9619047619047618
Stored 'accuracy_kfold_iris' (list)


### The average accuracy using Kfold validation for Iris Dataset is 96.2%

# Heart Data

In [45]:
df_heart = pd.read_csv("heart_disease_cleaned.csv")

In [46]:
df_heart.head()

age  sex  cp    trestbps        chol  fbs  restecg     thalach  exang  \
0   76    0   2  140.102822  197.105970    0      2.0  115.952071      0   
1   43    0   0  132.079599  341.049462    1      0.0  135.970028      1   
2   47    1   2  107.899290  242.822816    0      1.0  152.210039      0   
3   57    1   0  110.103508  334.952353    0      1.0  143.099327      1   
4   63    0   2  134.926653  252.023252    0      0.0  171.840974      0   

    oldpeak  slope  ca      thal  target  
0  1.284822    1.0   0  2.175904       1  
1  3.110483    1.0   0  3.082071       0  
2 -0.023723    2.0   0  2.020827       0  
3  3.082052    1.0   1  2.831509       0  
4  0.106792    2.0   0  1.903701       1

In [47]:
#Dividing the data into a training set and a test set (80%, 20%)
ip_train,ip_test,op_train,op_test = train_test_split(df_heart.iloc[:,:-1],df_heart.iloc[:,-1],test_size=0.2,random_state=98)

In [48]:
# Performing KNN by using default parameters on heart data
knn_heart = KNeighborsClassifier()

In [49]:
knn_heart.fit(ip_train,op_train)

KNeighborsClassifier()

In [50]:
heart_pred = knn_heart.predict(ip_test)

In [51]:
heart_prob = knn_heart.predict_proba(ip_test)

In [52]:
heart_accuracy = accuracy_score(heart_pred,op_test)

In [53]:
heart_accuracy

0.5757575757575758

In [54]:
print(classification_report(op_test,heart_pred))

              precision    recall  f1-score   support

           0       0.67      0.60      0.63        20
           1       0.47      0.54      0.50        13

    accuracy                           0.58        33
   macro avg       0.57      0.57      0.57        33
weighted avg       0.59      0.58      0.58        33



In [55]:
print(confusion_matrix(op_test,heart_pred))

[[12  8]
 [ 6  7]]


### The accuracy for heartdisease data set using default parameters is 57.57 %

# Validation for Heart Disease Dataset

In [56]:
k_train,k_validate,l_train,l_validate=train_test_split(ip_train,op_train,test_size=0.1,random_state=98)

In [57]:
k_validate.head()

age  sex  cp    trestbps        chol  fbs  restecg     thalach  exang  \
74    68    1   2  117.869131  276.905655    0      1.0  151.025569      0   
143   45    1   1  128.011535  308.060914    0      0.0  170.112376      0   
35    63    1   0  130.024365  254.108514    0      0.0  146.766086      0   
4     63    0   2  134.926653  252.023252    0      0.0  171.840974      0   
108   55    0   1  132.065557  342.106509    0      1.0  165.989654      0   
78    44    1   0  119.901676  168.945695    0      1.0  144.130286      1   
6     57    1   0  140.080577  192.215185    0      1.0  148.025188      0   
97    51    0   2  119.967348  294.961557    0      0.0  157.077183      0   
12    51    1   0  139.898949  298.081012    0      1.0  122.028452      1   
163   41    0   2  112.075764  268.005496    0      0.0  172.008896      1   
136   53    1   0  122.903381  282.053583    0      1.0   94.987305      1   
159   49    0   0  129.980215  268.817648    0      1.0  162.954304      0   
9     44    0   2  108.110824  140.910752    0      1.0  174.932949      0   
36    62    0   0  150.026813  243.936478    0      1.0  153.893137      1   

      oldpeak  slope  ca      thal  
74   0.855891    2.0   1  3.090707  
143 -0.103404    2.0   0  2.058228  
35   1.531057    1.0   1  3.026851  
4    0.106792    2.0   0  1.903701  
108  1.214041    2.0   0  1.913880  
78   2.978041    0.0   0  0.940130  
6    0.436121    1.0   0  0.931512  
97   0.458861    2.0   0  1.984220  
12   4.277138    1.0   3  2.884772  
163  0.085251    2.0   0  2.194162  
136  2.080513    1.0   2  3.148497  
159  0.198000    2.0   0  1.938370  
9    0.497061    1.0   0  2.281661  
36   1.418758    1.0   0  2.010602

In [64]:
accuracy_heart = []
for k in range(1,36):
    kn = KNeighborsClassifier(n_neighbors=k)
    kn.fit(k_train,l_train)
    accuracy_heart.append(accuracy_score(kn.predict(k_validate),l_validate))
    print("K = ",k,':: accuracy_heart =',accuracy_heart[k-1])
%store accuracy_heart knn_heart ip_test op_test ip_train op_train heart_pred heart_prob

K =  1 :: accuracy_heart = 0.5
K =  2 :: accuracy_heart = 0.35714285714285715
K =  3 :: accuracy_heart = 0.5
K =  4 :: accuracy_heart = 0.5714285714285714
K =  5 :: accuracy_heart = 0.6428571428571429
K =  6 :: accuracy_heart = 0.5714285714285714
K =  7 :: accuracy_heart = 0.7142857142857143
K =  8 :: accuracy_heart = 0.7142857142857143
K =  9 :: accuracy_heart = 0.7857142857142857
K =  10 :: accuracy_heart = 0.7857142857142857
K =  11 :: accuracy_heart = 0.9285714285714286
K =  12 :: accuracy_heart = 0.7142857142857143
K =  13 :: accuracy_heart = 0.7142857142857143
K =  14 :: accuracy_heart = 0.6428571428571429
K =  15 :: accuracy_heart = 0.5714285714285714
K =  16 :: accuracy_heart = 0.5714285714285714
K =  17 :: accuracy_heart = 0.5714285714285714
K =  18 :: accuracy_heart = 0.5714285714285714
K =  19 :: accuracy_heart = 0.7142857142857143
K =  20 :: accuracy_heart = 0.6428571428571429
K =  21 :: accuracy_heart = 0.6428571428571429
K =  22 :: accuracy_heart = 0.6428571428571429
K = 